src/test/java/spoon/test/filters/FilterTest.java

```diff
@@ -1124,4 +1125,17 @@ public void accept(CtType<?> ctType) {
 		// only one subtype remains unvisited
 		assertEquals(1, c2.counter);
 	}
+
+	@Test
+	public void testNameFilterWithGenericType() {
+		// contract: NameFilter of T should only return T elements
+
+		Launcher spoon = new Launcher();
+		spoon.addInputResource("./src/test/java/spoon/test/imports/testclasses/internal4/Constants.java");
+		spoon.buildModel();
+
+		CtType type = spoon.getFactory().Type().get(Constants.class);
+		List<CtMethod> ctMethods = type.getElements(new NameFilter<CtMethod>("CONSTANT"));
```

Comment:
> So it is not a bug of query engine, but old behavior of namefilter.

I agree but then usage of generic type parameter for NameFilter results is dangerous, as sometimes the result type are not expected. I sometimes get a `ClassCastException` when iterating over a result list of a query with NameFilter, because the type is not checked. 

So the new constructor is a first step, but maybe we should simply get rid of the type parameter for the NameFilter and explicitely return a collection of CtNamedElement: it will be clear that it does not check types. WDYT?